### ISSUE TITLE
VMMDLL MemReadScatter issue with data > 128 bytes

### ISSUE THREAD

**Issue by false** [ufrisk/MemProcFS#15]
Hello,

I am having an issue with MemReadScatter results, the pb data seems corrupted (/might be not read) over 128 bytes (at least in my tests it seems to be over 128). The size I am trying to read is initially around 4080 bytes, read result was not valid, so I've tried with 720 and I have realized first part of the data seems to be fine (around 128B) and all the rest is like not read.
With the function ReadEx on the same address same size to read, the bytes are the good ones.
Here are the results between use of both functions, used with a scatter call on only one address which is the same used on readex, page aligned :
Scatter :

> 0000    32 3e dc bd aa b7 5a 3f  1e 28 02 bf d6 5e 04 c3   2>....Z?.(...^..
0010    66 27 c5 bd 8e 02 00 3f  9e 52 5c 3f 98 9b df c3   f'.....?.R\?....
0020    c7 51 7d 3f b6 e4 10 3e  b8 3e e9 3c 89 f2 18 42   .Q}?...>.>.<...B
0030    d4 23 8d 3e bf 14 76 bf  00 00 00 00 00 00 00 c3   .#.>..v.........
0040    bf 14 76 3f d4 23 8d 3e  00 00 00 00 00 00 e0 c3   ..v?.#.>........
0050    00 00 00 00 00 00 00 00  00 00 80 3f 00 00 f0 be   ...........?....
0060    d4 23 8d 3e bf 14 76 bf  00 00 00 00 00 00 00 c3   .#.>..v.........
0070    bf 14 76 3f d4 23 8d 3e  00 00 00 00 00 00 e0 c3   ..v?.#.>........
0080    90 00 91 00 92 00 93 00  94 00 95 00 96 00 97 00   ................
0090    98 00 99 00 9a 00 9b 00  9c 00 9d 00 9e 00 9f 00   ................
00a0    a0 00 a1 00 a2 00 a3 00  a4 00 a5 00 a6 00 a7 00   ................
00b0    a8 00 a9 00 aa 00 ab 00  ac 00 ad 00 ae 00 af 00   ................
00c0    b0 00 b1 00 b2 00 b3 00  b4 00 b5 00 b6 00 b7 00   ................
00d0    b8 00 b9 00 ba 00 bb 00  bc 00 bd 00 be 00 bf 00   ................
00e0    c0 00 c1 00 c2 00 c3 00  c4 00 c5 00 c6 00 c7 00   ................
00f0    c8 00 c9 00 ca 00 cb 00  cc 00 cd 00 ce 00 cf 00   ................
0100    d0 00 d1 00 d2 00 d3 00  d4 00 d5 00 d6 00 d7 00   ................
0110    d8 00 d9 00 da 00 db 00  dc 00 dd 00 de 00 df 00   ................
0120    e0 00 e1 00 e2 00 e3 00  e4 00 e5 00 e6 00 e7 00   ................
0130    e8 00 e9 00 ea 00 eb 00  ec 00 ed 00 ee 00 ef 00   ................
0140    f0 00 f1 00 f2 00 f3 00  f4 00 f5 00 f6 00 f7 00   ................
0150    f8 00 f9 00 fa 00 fb 00  fc 00 fd 00 fe 00 ff 00   ................
0160    00 01 01 01 02 01 03 01  04 01 05 01 06 01 07 01   ................
0170    08 01 09 01 0a 01 0b 01  0c 01 0d 01 0e 01 0f 01   ................
0180    10 01 11 01 12 01 13 01  14 01 15 01 16 01 17 01   ................
0190    18 01 19 01 1a 01 1b 01  1c 01 1d 01 1e 01 1f 01   ................
01a0    20 01 21 01 22 01 23 01  24 01 25 01 26 01 27 01    .!.".#.$.%.&.'.
01b0    28 01 29 01 2a 01 2b 01  2c 01 2d 01 2e 01 2f 01   (.).*.+.,.-.../.
01c0    30 01 31 01 32 01 33 01  34 01 35 01 36 01 37 01   0.1.2.3.4.5.6.7.
01d0    38 01 39 01 3a 01 3b 01  3c 01 3d 01 3e 01 3f 01   8.9.:.;.<.=.>.?.
01e0    40 01 41 01 42 01 43 01  44 01 45 01 46 01 47 01   @.A.B.C.D.E.F.G.
01f0    48 01 49 01 4a 01 4b 01  4c 01 4d 01 4e 01 4f 01   H.I.J.K.L.M.N.O.
0200    50 01 51 01 52 01 53 01  54 01 55 01 56 01 57 01   P.Q.R.S.T.U.V.W.
0210    58 01 59 01 5a 01 5b 01  5c 01 5d 01 5e 01 5f 01   X.Y.Z.[.\.].^._.
0220    60 01 61 01 62 01 63 01  64 01 65 01 66 01 67 01   `.a.b.c.d.e.f.g.
0230    68 01 69 01 6a 01 6b 01  6c 01 6d 01 6e 01 6f 01   h.i.j.k.l.m.n.o.
0240    70 01 71 01 72 01 73 01  74 01 75 01 76 01 77 01   p.q.r.s.t.u.v.w.
0250    78 01 79 01 7a 01 7b 01  7c 01 7d 01 7e 01 7f 01   x.y.z.{.|.}.~...
0260    80 01 81 01 82 01 83 01  84 01 85 01 86 01 87 01   ................
0270    88 01 89 01 8a 01 8b 01  8c 01 8d 01 8e 01 8f 01   ................
0280    90 01 91 01 92 01 93 01  94 01 95 01 96 01 97 01   ................
0290    98 01 99 01 9a 01 9b 01  9c 01 9d 01 9e 01 9f 01   ................
02a0    a0 01 a1 01 a2 01 a3 01  a4 01 a5 01 a6 01 a7 01   ................
02b0    a8 01 a9 01 aa 01 ab 01  ac 01 ad 01 ae 01 af 01   ................
02c0    b0 01 b1 01 b2 01 b3 01  b4 01 b5 01 b6 01 b7 01   ................

ReadEx :

> 0000    32 3e dc bd aa b7 5a 3f  1e 28 02 bf d6 5e 04 c3   2>....Z?.(...^..
0010    66 27 c5 bd 8e 02 00 3f  9e 52 5c 3f 98 9b df c3   f'.....?.R\?....
0020    c7 51 7d 3f b6 e4 10 3e  b8 3e e9 3c 89 f2 18 42   .Q}?...>.>.<...B
0030    d4 23 8d 3e bf 14 76 bf  00 00 00 00 00 00 00 c3   .#.>..v.........
0040    bf 14 76 3f d4 23 8d 3e  00 00 00 00 00 00 e0 c3   ..v?.#.>........
0050    00 00 00 00 00 00 00 00  00 00 80 3f 00 00 f0 be   ...........?....
0060    d4 23 8d 3e bf 14 76 bf  00 00 00 00 00 00 00 c3   .#.>..v.........
0070    bf 14 76 3f d4 23 8d 3e  00 00 00 00 00 00 e0 c3   ..v?.#.>........
0080    00 00 00 00 00 00 00 00  00 00 80 3f 00 20 7e 42   ...........?. ~B
0090    e2 2c 00 3e ab 7d 58 3f  b4 d1 04 bf 3a 5c 04 c3   .,.>.}X?....:\..
00a0    54 a4 9f 3c 49 4f 05 3f  7a 7e 5a 3f 44 b2 df c3   T..<IO.?z~Z?D...
00b0    09 f0 7d 3f f8 7f ef bd  37 7b 47 3d fd e2 1c 42   ..}?....7{G=...B
00c0    ac 42 46 3e ee b9 43 3f  fc 62 1d bf f0 df 03 c3   .BF>..C?.b......
00d0    e0 42 20 3d 0d c1 1e 3f  ca 94 48 3f 97 a8 df c3   .B =...?..H?....
00e0    c6 f4 7a 3f 12 f9 33 be  7b a1 b8 3d cb 46 2c 42   ..z?..3.{..=.F,B
00f0    b8 59 7d 3e 3e fe 2e 3f  fe c9 2f bf 7c f6 02 c3   .Y}>>..?../.|...
0100    fc 86 04 3c dd 0c 35 3f  02 fa 34 3f 00 91 df c3   ...<..5?..4?....
0110    4b 08 78 3f 93 ca 38 be  9e 83 2d 3e d0 be 3e 42   K.x?..8...->..>B
0120    b8 8d 9c 3e 0d e7 06 3f  af 00 4b bf b3 3f 01 c3   ...>...?..K..?..
0130    1b e5 8c 3e ba 97 3f 3f  fd 7b 1a 3f d4 89 df c3   ...>..??.{.?....
0140    22 56 69 3f 20 33 ce be  96 ad ab 3d fe 97 59 42   "Vi? 3.....=..YB
0150    7e 9f b6 3e a2 8b eb 3e  44 27 50 bf 17 b1 fe c2   ~..>...>D'P.....
0160    9a 21 07 3f 88 6d 1e 3f  ab ea 14 3f 9d ae de c3   .!.?.m.?...?....
0170    96 53 45 3f d3 fd 22 bf  48 10 b5 bc b4 48 70 42   .SE?..".H....HpB
0180    e0 eb ad 3d 2e fd 80 3e  7e c9 76 bf 37 c2 fb c2   ...=...>~.v.7...
0190    00 e6 37 3e 7a a2 72 3f  2c eb 86 3e cf 1d dd c3   ..7>z.r?,..>....
01a0    d4 e5 7a 3f 1e 32 48 be  bc 5c 10 3d 15 46 7f 42   ..z?.2H..\.=.F.B
01b0    d7 b1 49 3f 54 58 f6 3d  76 9e 1a bf be bd 00 c3   ..I?TX.=v.......
01c0    9e da 1c bf 16 f5 80 3e  d8 c4 3f bf df 9a df c3   .......>..?.....
01d0    6b 07 7e 3d 60 d3 75 3f  12 56 8b 3e 62 3a 7d 42   k.~=`.u?.V.>b:}B
01e0    76 d8 71 bf 4e 15 a7 3e  50 89 03 3d 65 b5 fe c2   v.q.N..>P..=e...
01f0    95 f8 a5 3e 2c b6 71 3f  50 f5 6f bd ed 6f dd c3   ...>,.q?P.o..o..
0200    98 80 4a bd 48 0d 38 bd  99 6d 7f bf d5 e5 64 42   ..J.H.8..m....dB
0210    18 b1 3a be 02 30 46 3f  e3 2c 1b 3f b4 a6 05 c3   ..:..0F?.,.?....
0220    a3 4d 4b 3f 97 77 f5 3e  9c 35 bf be 5c 5b dc c3   .MK?.w.>.5..\[..
0230    e0 68 14 bf 05 9b d3 3e  7e c4 33 bf 62 94 63 42   .h.....>~.3.b.cB
0240    be 51 0a 3f 25 da 1a 3f  a3 bf 15 3f f4 c3 07 c3   .Q.?%..?...?....
0250    a5 66 57 3f d1 f8 ca be  47 08 bc be 8a c0 d7 c3   .fW?....G.......
0260    40 b2 1f 3c 3c cc 30 3f  bf 20 39 bf 28 b0 48 42   @..<<.0?. 9.(.HB
0270    ac 84 19 3f 47 de f6 3e  f6 7f 23 bf 16 8f 01 c3   ...?G..>..#.....
0280    71 0d 49 3f 78 41 4e be  a1 d9 15 3f 63 eb d2 c3   q.I?xAN....?c...
0290    ba 47 1d 3e c3 44 5a bf  7e b9 ff be cf 22 49 42   .G.>.DZ.~...."IB
02a0    45 93 d7 3e 90 16 18 3f  63 76 2f bf 74 ec 00 c3   E..>...?cv/.t...
02b0    65 26 54 3f 0c 2f 62 3d  08 94 0e 3f 65 ab d2 c3   e&T?./b=...?e...
02c0    c4 c9 bc 3e 80 70 4d bf  82 2b f0 be 35 07 4a 42   ...>.pM..+..5.JB

**Comment by ufrisk** [comment]
Can you please check the bytes read in the cb field of the MEM_IO_SCATTER_HEADER struct? It should be 0x1000 bytes on a successful read.

**Comment by false** [comment]
Ok, that's some strange bug : After a full a reboot on a 3840 bytes it now reads 1216 bytes successfully, rest is not valid. ReadEx reads the 3840 bytes successfully.
Readscatter ppMem cb is 4096, cbmax too.

**Comment by false** [comment]
Edit : reboot does not have an actual effet, it seems it depends on the addresses I'm accessing. I have try to read every float values by specifying the 240 desired addresses using read scatter but I get the same results : First ones are valid, next are not, readex reads everythings good.

**Comment by ufrisk** [comment]
This is quite intriguing since ReadEx is just an wrapper around ReadScatter. Every call to ReadEx is sent onward to ReadScatter. In the case if you're doing page aligned reads it should collapse down to the exact same case - i.e. one single ReadScatter page read.

Maybe if you copy the function below and replace the pProcess object with the PID you may be able to try to locate the error? If the error is in your code or in mine.

Or do you mean that you get the error in page 2+ onwards in a non-contiguous memory read to ReadScatter (i.e. never in 1st page or when only reading 1 page)?

https://github.com/ufrisk/MemProcFS/blob/69c2a5767e9beee84727ce75cb93dcbd96fa1f01/vmm/vmm.c#L1387

**Comment by false** [comment]
I have been trying to locate the issue but didn't really find it unfortunatly. Thus I have made a very small application exemple, and read calls which would let you reproduce you the problem. Can I send the source code to your email ?

**Comment by ufrisk** [comment]
Please drop me an email at pcileech@frizk.net and I'll see if I can find something obvious. If I have to look a bit deeper into things it's going to be a few weeks from now though.

**Comment by ufrisk** [comment]
Thank you for reporting this issue, and apologies for having to wait quite a while.

I was quite intrigued by this since this function is the most central one in MemProcFS/PCILeech/LeechCore and if it would send corrupt data everything would fall apart.

It seems that if you scatter items that pass a page boundary (0x1000) it will fail though. I never had any need for it so I never really tested that.

It's a lot of work and testing for me to look into this and allow this to support reading cross page boundaries (and I have no need for it internally, neither do I think it's desirable since it would drive code complexity) so I'll "fix" this by updating the documentation in the header files.

Apologies for you getting bitten by this. And thank you for your excellent test program.

**Comment by false** [comment]
Back to this issue. Would it be possible to get some kind of error code when the read failed then ? I switched to MemReadScatter calls only and often get corrupted reads, that I don't know about. My use case is building up big arrays of addresses and I then make one MemReadScatter call (then treat results).

( Note. It is not very clear the advantage of using MemReadScatter if it gets the same 0x1000 page boundary limit than a basic readEx : I though you could provide x addresses to MemReadScatter and it would dispatch it in a optimzed number of calls whatever the size or address you access to ; But basically if it is the same 0x1000 limit, then it does not really differ from directly using readEx ? )

**Comment by ufrisk** [comment]
I've added it to the API documentation that reads in single scatter entries should not cross page boundaries. I hope that should be sufficient.

Anyway, I think it's best and most safe to look at the ReadScatter function as X * ReadPage(). That is that you'll be able to read an arbitrary amount of non contiguous memory pages.

ReadEx reads one contiguous range of memory per call.

If you're about to read memory all over the place in different memory pages then ReadScatter would be more efficient - but it would add extra work on your side to use it.

**Comment by ufrisk** [comment]
I'm closing this issue since it should now be documented that individual scatter reads should not cross a page boundary.

**Comment by SetWait** [comment]
I was unable to read the content while using it. The function returned the normal number of items read, but the buffer content was 0. I used CE to check that the address clearly had data
![1685334626957
![1685334541786](https://github.com/ufrisk/MemProcFS/assets/60318979/78680505-b991-4b7d-a934-e265c5ea1dc1)
](https://github.com/ufrisk/MemProcFS/assets/60318979/58fbc2ab-58d7-44c8-a03c-ae30b7770886)

**Comment by SetWait** [comment]
![1685334541786](https://github.com/ufrisk/MemProcFS/assets/60318979/5976d601-b8c0-4817-9296-3933a325ef76)

**Comment by ufrisk** [comment]
@SetWait The original "older" scatter function you use is quite complicated. It have a lot of rules with sizes, alignment and that it's forvodden to cross a page boundary (i.e. 0x1000).

It's probably better to use the more recent Scatter API which is made to be much more user friendly. The new API shouldn't have any restrictions like this. Have a look at `VMMDLL_Scatter_Initialize()` and following functions. There should also be some examples around it.

Also I've deleted some of the earlier posts since I don't wish for people to discuss cheats with screenshots in my project. It was not why I created it and it's not its primary use. General questions are perfectly fine though. Thanks for understanding.
